fix(compiler): scope nested CSS rules

crisbeto opened the pull request:
[Browsers now support nesting CSS rules](https://developer.chrome.com/articles/css-nesting/) however our emulated encapsulation wasn't processing rules nested inside of other rules, except for at-rules.

These changes update the encapsulation logic to process rules inside of other rules. I also had to fix a couple of additional bugs where we didn't handle the ampersand symbol in a selector correctly and we weren't skipping style blocks inside of strings.

crisbeto commented on packages/compiler/test/shadow_css/keyframes_spec.ts:
```diff
@@ -395,14 +395,14 @@ describe('ShadowCss, keyframes and animations', () => {
 
         @keyframes foo {}
         @keyframes 'fo\\'o' {}
-        @keyframes 'foo'' {}
```
I don't know the context of this assertion, however this is actually invalid syntax. I had to update the test, because after my changes, invalid syntax would prevent styles from being encapsulated correctly.

crisbeto commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -842,7 +855,7 @@ const _cssSemiInPlaceholderReGlobal = new RegExp(SEMI_IN_PLACEHOLDER, 'g');
 
 export class CssRule {
-  constructor(public selector: string, public content: string) {}
+  constructor(public selector: string, public content: string, readonly isBlock: boolean) {}
```
The context behind the `isBlock` flag is that as the way the parser is set up right now, it doesn't distinguish between something like `color: red` as a CSS property and `color:red` as a selector. Until now it wasn't a problem since we didn't scope anything inside of other rules (except for `@media`, `@container` etc).

pkozlowski-opensource commented:
@crisbeto this change looks reasonable to me. As discussed yesterday, we should run TGP on it. 

I also wonder if you want to have another pair of eyes on this PR, from someone more familiar with the CSS syntax / this parser bit.

crisbeto commented:
The TGP showed one legit failure in a file with a bunch of base64-encoded images. I'll have to investigate deeper.

crisbeto committed:
fix(compiler): scope nested CSS rules

[Browsers now support nesting CSS rules](https://developer.chrome.com/articles/css-nesting/) however our emulated encapsulation wasn't processing rules nested inside of other rules, except for at-rules.

These changes update the encapsulation logic to process rules inside of other rules. I also had to fix a couple of additional bugs where we didn't handle the ampersand symbol in a selector correctly and we weren't skipping style blocks inside of strings.
